crates/sc-consensus-subspace-rpc/src/lib.rs

```diff
@@ -304,6 +318,60 @@ where
             _block: PhantomData,
         })
     }
+
+    /// Returns the last archived segment if available.
+    /// If there is no archived segment, reconstructs the genesis segment if it contains `requested_piece_index`.
+    ///
+    /// If there is a last archived segment, `requested_piece_index` is ignored.
+    pub fn last_archived_segment(
+        &self,
+        requested_piece_index: PieceIndex,
+    ) -> Result<Option<Arc<NewArchivedSegment>>, Error> {
+        let archived_segment = {
+            let mut cached_archived_segment = self.cached_archived_segment.lock();
+
+            match cached_archived_segment
+                .as_ref()
+                .and_then(CachedArchivedSegment::get)
+            {
+                Some(archived_segment) => archived_segment,
+                None => {
+                    if requested_piece_index > SegmentIndex::ZERO.last_piece_index() {
+                        // TODO:
+                        // - fall back to updating the cache
+                        // - split out the cache update code from subscribe_archived_segment_header() into its own spawned task,
+                        //   which keeps the cache up-to-date even when there isn't an active subscription
```

Comment:
This cache only exists for a single purpose of serving pieces to farmers. It is useless for general RPC use since only weak reference is intentionally persisted, so any use of it outside of the use case of pulling pieces for farming purposes will result in `Ok(None)`.

It should not be used for object mappings and should not be needed for object mappings either.